### PR TITLE
td-shim: remove the use of `Vec` in `e820`

### PR DESCRIPTION
**Pull request by gaojiaqi7**
Reducing the use of heap can improve the functional safety. For e820 implementation, use an fixed `MAX_E820_ENTRY` size array to replace the vector.

Relate issue: https://github.com/confidential-containers/td-shim/issues/685

### CODE REVIEW

File: td-shim/src/bin/td-shim/e820.rs [hunk 1]
```diff
@@ -158,27 +154,58 @@ impl E820Table {
...
+    fn insert_entry(&mut self, index: usize, entry: E820Entry) -> Result<(), E820Error> {
+        if index >= MAX_E820_ENTRY {
+            return Err(E820Error::InvalidIndex);
+        }
+        if self.num_entries >= MAX_E820_ENTRY {
+            return Err(E820Error::TooManyEntries);
+        }
+
+        // Move all entries after the index back one position
+        for idx in (index + 1..self.num_entries + 1).rev() {
```
Comment (jyao1) [inline td-shim/src/bin/td-shim/e820.rs:183]: what if the index > self.num_entries?

File: td-shim/src/bin/td-shim/e820.rs [hunk 2]
```diff
@@ -158,27 +154,58 @@ impl E820Table {
...
+        Ok(())
+    }
+
+    fn remove_entry(&mut self, index: usize) -> Result<(), E820Error> {
+        if index >= MAX_E820_ENTRY {
+            return Err(E820Error::InvalidIndex);
+        }
+
+        // Move all entries after the index forward one position
+        for idx in index + 1..self.num_entries {
```
Comment (jyao1) [inline td-shim/src/bin/td-shim/e820.rs:198]: what if the index > self.num_entries?

File: td-shim/src/bin/td-shim/e820.rs [hunk 3]
```diff
@@ -158,27 +154,58 @@ impl E820Table {
...
+            self.entries[idx + 1] = self.entries[idx];
+        }
+
+        self.entries[index] = entry;
+        self.num_entries += 1;
+        Ok(())
+    }
+
+    fn remove_entry(&mut self, index: usize) -> Result<(), E820Error> {
+        if index >= self.num_entries {
```
Comment (jyao1) [inline td-shim/src/bin/td-shim/e820.rs:193]: What if self.num_entries is 0 ?

Comment (gaojiaqi7) [reply td-shim/src/bin/td-shim/e820.rs:193]: if `self.num_entries == 0`, the `index >= self.num_entries` will always be true, then it will return error

**Review by jyao1 — APPROVED** [review]

**Pull request merged**